### PR TITLE
Expose refresh indicators for kernels

### PR DESCRIPTION
**Pull request by DonJayamanne**

### CODE REVIEW

File: src/kernels/raw/finder/localKernelSpecFinderBase.node.ts [hunk 1]
```diff
@@ -80,31 +80,39 @@ export class LocalKernelSpecFinder implements IDisposable {
         const key = getComparisonKey(specPath);
         // If we have not already loaded this kernel spec, then load it
         if (!this.pathToKernelSpec.has(key)) {
-            this.pathToKernelSpec.set(key, this.loadKernelSpec(specPath, cancelToken, interpreter));
+            const promise = this.loadKernelSpec(specPath, cancelToken, interpreter).then((kernelSpec) => {
```
Comment (DonJayamanne) [inline src/kernels/raw/finder/localKernelSpecFinderBase.node.ts:83]: More readable

File: src/kernels/raw/finder/localPythonAndRelatedNonPythonKernelSpecFinder.node.ts [hunk 2]
```diff
@@ -315,15 +315,15 @@ export class LocalPythonAndRelatedNonPythonKernelSpecFinder
                 this.interpreterService.getActiveInterpreter(folder.uri)
             )
         );
-        const [kernelSpecs, activeInterpreters, globalKernelSpecs, tempDirForKernelSpecs] = await Promise.all([
+        const [kernelSpecs, activeInterpreters, tempDirForKernelSpecs] = await Promise.all([
             this.findKernelSpecsInInterpreters(interpreters, cancelToken),
             activeInterpreterInAWorkspacePromise,
-            this.listGlobalPythonKernelSpecsIncludingThoseRegisteredByUs(),
```
Comment (DonJayamanne) [inline src/kernels/raw/finder/localPythonAndRelatedNonPythonKernelSpecFinder.node.ts:321]: This was not async, it was a sync function

File: src/kernels/raw/finder/localPythonAndRelatedNonPythonKernelSpecFinder.node.ts [hunk 3]
```diff
@@ -579,10 +579,11 @@ export class LocalPythonAndRelatedNonPythonKernelSpecFinder
         });
     }
 
-    private async findMatchingInterpreter(
-        kernelSpec: IJupyterKernelSpec,
-        interpreters: PythonEnvironment[]
-    ): Promise<PythonEnvironment | undefined> {
+    private async findMatchingInterpreter(kernelSpec: IJupyterKernelSpec): Promise<PythonEnvironment | undefined> {
+        const interpreters = this.extensionChecker.isPythonExtensionInstalled
```
Comment (DonJayamanne) [inline src/kernels/raw/finder/localPythonAndRelatedNonPythonKernelSpecFinder.node.ts:583]: No need of passing an argument.

File: src/kernels/raw/finder/contributedLocalKernelSpecFinder.node.ts [hunk 4]
```diff
@@ -75,7 +75,48 @@ export class ContributedLocalKernelSpecFinder
...
+                nonPythonKernelDiscoveryPromise = createDeferred<void>();
+                this.promiseMonitor.push(nonPythonKernelDiscoveryPromise.promise);
+            }
+        });
+        let pythonKernelDiscoveryPromise: Deferred<void> | undefined = undefined;
+        if (this.pythonKernelFinder.status === 'discovering') {
+            pythonKernelDiscoveryPromise = createDeferred<void>();
+            this.promiseMonitor.push(pythonKernelDiscoveryPromise.promise);
+        }
+        this.pythonKernelFinder.onDidChangeStatus(() => {
```
Comment (DonJayamanne) [inline src/kernels/raw/finder/contributedLocalKernelSpecFinder.node.ts:109]: Combine status of the 3 sources into one.

File: src/platform/api/pythonApi.ts [hunk 5]
```diff
@@ -407,43 +407,13 @@ export class InterpreterService implements IInterpreterService {
...
             this.interpreterListCachePromise = this.getInterpretersImpl(cancellation.token);
-            this.interpreterListCachePromise.finally(() => cancellation.dispose);
+            this.interpreterListCachePromise.finally(() => cancellation.dispose());
             this.refreshPromises.push(this.interpreterListCachePromise);
         }
         return this.interpreterListCachePromise;
     }
-    private _waitForAllInterpretersToLoad?: Promise<void>;
     public async waitForAllInterpretersToLoad(): Promise<void> {
-        if (!this._waitForAllInterpretersToLoad) {
```
Comment (DonJayamanne) [inline src/platform/api/pythonApi.ts:417]: Simpler & identical to just wait for `getInterpreters`

File: src/notebooks/controllers/kernelSource/notebookKernelSourceSelector.ts [hunk 6]
```diff
@@ -648,33 +648,24 @@ export class NotebookKernelSourceSelector implements INotebookKernelSourceSelect
...
         });
-        // Assume we're always busy.
-        quickPick.busy = true;
-        let noLongerBusy = true;
-        // If we don't get any updates after 2s, then hide the busy indicator
-        let updateTimeout = setTimeout(() => {
-            if (noLongerBusy) {
-                quickPick.busy = false;
-            }
-        }, 2_000);
```
Comment (DonJayamanne) [inline src/notebooks/controllers/kernelSource/notebookKernelSourceSelector.ts:659]: No need to assume we're busy at the start

Comment (DonJayamanne) [reply src/notebooks/controllers/kernelSource/notebookKernelSourceSelector.ts:659]: We're better at detecting the states now

File: src/kernels/raw/finder/contributedLocalKernelSpecFinder.node.ts [hunk 7]
```diff
@@ -75,28 +75,37 @@ export class ContributedLocalKernelSpecFinder
...
-            if (this.interpreters.status === 'idle') {
-                deferred?.resolve();
-                deferred = undefined;
-            } else if (this.interpreters.status === 'refreshing') {
-                if (deferred) {
-                    return;
+
+        this.loadData().then(noop, noop);
+        let combinedProgress: Deferred<void> | undefined = undefined;
+        const updateCombinedStatus = () => {
```
Comment (DonJayamanne) [inline src/kernels/raw/finder/contributedLocalKernelSpecFinder.node.ts:88]: Previously we only checked status of one, now we monitor status of 3 items (resulting in more accurate status reports)

File: src/kernels/raw/finder/contributedLocalPythonEnvFinder.node.ts [hunk 8]
```diff
@@ -69,26 +69,32 @@ export class ContributedLocalPythonEnvFinder
...
-        }
-        this.interpreters.onDidChangeStatus(() => {
-            if (this.interpreters.status === 'idle') {
-                deferred?.resolve();
-                deferred = undefined;
-            } else if (this.interpreters.status === 'refreshing') {
-                if (deferred) {
-                    return;
+        let combinedProgress: Deferred<void> | undefined = undefined;
+        const updateCombinedStatus = () => {
```
Comment (DonJayamanne) [inline src/kernels/raw/finder/contributedLocalPythonEnvFinder.node.ts:80]: Previously we only checked status of one, now we monitor status of 2 items (resulting in more accurate status reports)

**Review by TylerLeonhardt — APPROVED** [review]

**Pull request merged**